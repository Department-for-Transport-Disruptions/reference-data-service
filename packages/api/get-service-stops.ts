import { APIGatewayEvent, APIGatewayProxyResultV2 } from "aws-lambda";
import {
    getServiceStops,
    isValidBusStopType,
    isValidMode,
    ServiceStop,
    ServiceStops,
    ServiceStopsQueryInput,
    ServiceTracks,
    Stops,
} from "./client";
import { ClientError } from "./error";
import { executeClient } from "./execute-client";
import { isServiceStops } from "./utils";

export const main = async (event: APIGatewayEvent): Promise<APIGatewayProxyResultV2> =>
    executeClient(event, getQueryInput, getServiceStops, formatStops);

const MAX_ADMIN_AREA_CODES = process.env.MAX_ADMIN_AREA_CODES || "5";

export const getQueryInput = (event: APIGatewayEvent): ServiceStopsQueryInput => {
    const { pathParameters } = event;

    const serviceId = pathParameters?.serviceId;

    if (!serviceId) {
        throw new ClientError("Service ID must be provided");
    }

    if (isNaN(Number(serviceId))) {
        throw new ClientError("Provided service ID is not valid");
    }

    const stopTypes = pathParameters.stopTypes || "";
    const stopTypesArray = stopTypes
        .split(",")
        .filter((stop) => stop)
        .map((stop) => stop.trim());

    const modes = pathParameters.modes || "";
    const modesArray = modes
        .split(",")
        .filter((mode) => mode)
        .map((mode) => mode.trim());

    const filteredModesArray = modesArray.filter(isValidMode);

    if (filteredModesArray.length !== modesArray.length) {
        throw new ClientError("Invalid mode provided");
    }

    const adminAreaCodes = pathParameters?.adminAreaCodes ?? "";
    const adminAreaCodeArray = adminAreaCodes
        .split(",")
        .filter((adminAreaCode) => adminAreaCode)
        .map((adminAreaCode) => adminAreaCode.trim());

    if (adminAreaCodeArray.length > Number(MAX_ADMIN_AREA_CODES)) {
        throw new ClientError(`Only up to ${MAX_ADMIN_AREA_CODES} administrative area codes can be provided`);
    }

    const busStopTypes = pathParameters?.busStopTypes || "";
    const busStopTypesArray = busStopTypes
        .split(",")
        .filter((stop) => stop)
        .map((busStopType) => busStopType.trim());

    const filteredBusStopTypesArray = busStopTypesArray.filter(isValidBusStopType);

    if (filteredBusStopTypesArray.length !== busStopTypesArray.length) {
        throw new ClientError("Invalid bus stop type provided");
    }

    return {
        serviceId: Number(serviceId),
        ...(filteredBusStopTypesArray && filteredBusStopTypesArray.length > 0
            ? { busStopTypes: filteredBusStopTypesArray }
            : {}),
        ...(filteredModesArray && filteredModesArray.length > 0 ? { modes: filteredModesArray } : {}),
        ...(stopTypesArray && stopTypesArray.length > 0 ? { stopTypes: stopTypesArray } : {}),
        ...(adminAreaCodes && adminAreaCodeArray.length > 0 ? { adminAreaCodes: adminAreaCodeArray } : {}),
    };
};

export const flattenStops = (stops: ServiceStops): ServiceStop[] => {
    return stops.flatMap((stop) => {
        const stopArray: ServiceStop[] = [];
        if (stop.fromStatus === "active") {
            stopArray.push({
                id: stop.fromId,
                atcoCode: stop.fromAtcoCode,
                naptanCode: stop.fromNaptanCode,
                commonName: stop.fromCommonName,
                street: stop.fromStreet,
                indicator: stop.fromIndicator,
                bearing: stop.fromBearing,
                nptgLocalityCode: stop.fromNptgLocalityCode,
                localityName: stop.fromLocalityName,
                parentLocalityName: stop.fromParentLocalityName,
                longitude: stop.fromLongitude,
                latitude: stop.fromLatitude,
                stopType: stop.fromStopType,
                busStopType: stop.fromBusStopType,
                timingStatus: stop.fromTimingStatus,
                administrativeAreaCode: stop.fromAdministrativeAreaCode,
                status: stop.fromStatus,
                direction: stop.direction || "",
                sequenceNumber: stop.fromSequenceNumber || "",
                journeyPatternId: stop.journeyPatternId,
            });
        }

        if (stop.toStatus === "active") {
            stopArray.push({
                id: stop.toId,
                atcoCode: stop.toAtcoCode,
                naptanCode: stop.toNaptanCode,
                commonName: stop.toCommonName,
                street: stop.toStreet,
                indicator: stop.toIndicator,
                bearing: stop.toBearing,
                nptgLocalityCode: stop.toNptgLocalityCode,
                localityName: stop.toLocalityName,
                parentLocalityName: stop.toParentLocalityName,
                longitude: stop.toLongitude,
                latitude: stop.toLatitude,
                stopType: stop.toStopType,
                busStopType: stop.toBusStopType,
                timingStatus: stop.toTimingStatus,
                administrativeAreaCode: stop.toAdministrativeAreaCode,
                status: stop.toStatus,
                direction: stop.direction || "",
                sequenceNumber: stop.toSequenceNumber || "",
                journeyPatternId: stop.journeyPatternId,
            });
        }

        return stopArray;
    });
};

// eslint-disable-next-line @typescript-eslint/require-await
export const formatStops = async (stops: ServiceStops | ServiceTracks): Promise<Stops> => {
    if (!isServiceStops(stops)) {
        return [];
    }

    return flattenStops(stops).filter(
        (flattenedStop, index, self) => index === self.findIndex((stop) => stop.atcoCode === flattenedStop.atcoCode),
    );
};
