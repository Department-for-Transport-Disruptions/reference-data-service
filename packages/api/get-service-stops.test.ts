import { APIGatewayEvent } from "aws-lambda";
import { describe, expect, it } from "vitest";
import { formatStops, getQueryInput } from "./get-service-stops";
import { stopsDbData } from "./test/testdata";
import { VehicleMode, BusStopType } from "./client";

describe("get-service-stops", () => {
    describe("input generation", () => {
        it("handles serviceId", () => {
            const event = {
                pathParameters: {
                    serviceId: "234",
                },
            } as unknown as APIGatewayEvent;

            expect(getQueryInput(event)).toEqual({ serviceId: 234 });
        });

        it("handles serviceId, stopTypes, busStopType and modes", () => {
            const event = {
                pathParameters: {
                    serviceId: "234",
                    stopTypes: "BCT",
                    busStopTypes: BusStopType.MKD,
                    modes: VehicleMode.bus,
                },
            } as unknown as APIGatewayEvent;

            expect(getQueryInput(event)).toEqual({
                serviceId: 234,
                stopTypes: ["BCT"],
                busStopTypes: [BusStopType.MKD],
                modes: [VehicleMode.bus],
            });
        });

        it("throws a ClientError if no serviceId provided", () => {
            const event = {
                pathParameters: {},
            } as unknown as APIGatewayEvent;

            expect(() => getQueryInput(event)).toThrowError("Service ID must be provided");
        });

        it("throws a ClientError if invalid serviceId provided", () => {
            const event = {
                pathParameters: {
                    serviceId: "abc",
                },
            } as unknown as APIGatewayEvent;

            expect(() => getQueryInput(event)).toThrowError("Provided service ID is not valid");
        });
        it("throws a ClientError if invalid busStopType provided", () => {
            const event = {
                pathParameters: {
                    serviceId: "234",
                    busStopTypes: "invalid",
                },
            } as unknown as APIGatewayEvent;

            expect(() => getQueryInput(event)).toThrowError("Invalid bus stop type provided");
        });
    });

    describe("format service", () => {
        it("correctly formats db response", async () => {
            const formattedService = await formatStops(stopsDbData);

            expect(formattedService).toMatchSnapshot();
        });
    });
});
