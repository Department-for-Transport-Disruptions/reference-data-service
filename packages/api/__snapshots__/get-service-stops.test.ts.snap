// Vitest Snapshot v1

exports[`get-service-stops > format service > correctly formats db response 1`] = `
[
  {
    "administrativeAreaCode": "075",
    "atcoCode": "1000DBAR3758",
    "bearing": "N",
    "busStopType": "MKD",
    "commonName": "Old Road",
    "direction": "outbound",
    "id": 388498,
    "indicator": "Opp",
    "latitude": "53.354415",
    "localityName": "Bamford",
    "longitude": "-1.695986",
    "naptanCode": "dbsgwgmw",
    "nptgLocalityCode": "E0045121",
    "parentLocalityName": "",
    "sequenceNumber": "1",
    "status": "active",
    "stopType": "BCT",
    "street": "Ashopton Road",
    "timingStatus": "OTH",
  },
  {
    "administrativeAreaCode": "075",
    "atcoCode": "1000DBAR3762",
    "bearing": "N",
    "busStopType": "CUS",
    "commonName": "Lydgate Cottages",
    "direction": "outbound",
    "id": 388501,
    "indicator": "Opp",
    "latitude": "53.358629",
    "localityName": "Yorkshire Bridge",
    "longitude": "-1.698976",
    "naptanCode": "dbsgwgpg",
    "nptgLocalityCode": "N0064822",
    "parentLocalityName": "",
    "sequenceNumber": "2",
    "status": "active",
    "stopType": "BCT",
    "street": "Ashopton Road",
    "timingStatus": "OTH",
  },
  {
    "administrativeAreaCode": "075",
    "atcoCode": "1000DCSC3082",
    "bearing": "S",
    "busStopType": "MKD",
    "commonName": "Cliff College",
    "direction": "outbound",
    "id": 386465,
    "indicator": "Adj",
    "latitude": "53.261588",
    "localityName": "Calver",
    "longitude": "-1.629153",
    "naptanCode": "dbsgdpmp",
    "nptgLocalityCode": "E0045018",
    "parentLocalityName": "",
    "sequenceNumber": "3",
    "status": "active",
    "stopType": "BCT",
    "street": "Baslow Road",
    "timingStatus": "OTH",
  },
  {
    "administrativeAreaCode": "075",
    "atcoCode": "1000DCSB3083",
    "bearing": "S",
    "busStopType": "MKD",
    "commonName": "Stanton Ford",
    "direction": "outbound",
    "id": 386459,
    "indicator": "Adj",
    "latitude": "53.257401",
    "localityName": "Baslow",
    "longitude": "-1.626806",
    "naptanCode": "dbsgdpjt",
    "nptgLocalityCode": "E0006776",
    "parentLocalityName": "",
    "sequenceNumber": "4",
    "status": "active",
    "stopType": "BCT",
    "street": "Calver Road",
    "timingStatus": "TIP",
  },
  {
    "administrativeAreaCode": "075",
    "atcoCode": "1000DYBA3764",
    "bearing": "N",
    "busStopType": "MKD",
    "commonName": "Lydgate Lane",
    "direction": "inbound",
    "id": 381418,
    "indicator": "adj",
    "latitude": "53.360453",
    "localityName": "Yorkshire Bridge",
    "longitude": "-1.698678",
    "naptanCode": "dbsdgawa",
    "nptgLocalityCode": "N0064822",
    "parentLocalityName": "",
    "sequenceNumber": "4",
    "status": "active",
    "stopType": "BCT",
    "street": "Ashopton Road",
    "timingStatus": "OTH",
  },
  {
    "administrativeAreaCode": "075",
    "atcoCode": "1000DBCR3084",
    "bearing": "S",
    "busStopType": "MKD",
    "commonName": "White Lodge Lane",
    "direction": "inbound",
    "id": 387701,
    "indicator": "Adj",
    "latitude": "53.250479",
    "localityName": "Baslow",
    "longitude": "-1.626567",
    "naptanCode": "dbsjadtp",
    "nptgLocalityCode": "E0006776",
    "parentLocalityName": "",
    "sequenceNumber": "8",
    "status": "active",
    "stopType": "BCT",
    "street": "Calver Road",
    "timingStatus": "OTH",
  },
]
`;
